lib/core/src/Cardano/Pool/DB.hs

```diff
@@ -166,6 +190,98 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
         -- For a Sqlite DB, this would be "run a query inside a transaction".
     }
 
+-- | Represents an abstract notion of a certificate publication time.
+--
+-- Certificates published at later times take precedence over certificates
+-- published at earlier times.
+--
+type CertificatePublicationTime = (SlotId, SlotInternalIndex)
+
+-- | Indicates the current registration status of a pool.
+--
+-- Use the 'readPoolRegistrationStatus' function to query the registration
+-- status for a particular pool and database backend.
+--
+data PoolRegistrationStatus
+    = PoolNotRegistered
+        -- ^ Indicates that a pool is not registered.
+    | PoolRegistered
+        PoolRegistrationCertificate
+        -- ^ Indicates that a pool is registered BUT NOT marked for retirement.
+        -- Records the latest registration certificate.
+    | PoolRegisteredAndRetired
+        PoolRegistrationCertificate
+        PoolRetirementCertificate
+        -- ^ Indicates that a pool is registered AND ALSO marked for retirement.
+        -- Records the latest registration and retirement certificates.
+    deriving (Eq, Show)
+
+-- | Given the latest registration and retirement certificates for a pool,
+-- determine the pool's current registration status based on the relative
+-- order in which the certificates were published.
+determinePoolRegistrationStatus
+    :: Ord certificatePublicationTime
+    => Maybe (certificatePublicationTime, PoolRegistrationCertificate)
+    -> Maybe (certificatePublicationTime, PoolRetirementCertificate)
+    -> PoolRegistrationStatus
+determinePoolRegistrationStatus = f
+  where
+    f Nothing _ =
+        PoolNotRegistered
+    f (Just (_, regCert)) Nothing =
+        PoolRegistered regCert
+    f (Just reg) (Just ret) =
+        g reg ret
+
+    g (regTime, regCert) (retTime, retCert)
+        | regPoolId /= retPoolId =
+            differentPoolsError
+        | regTime > retTime =
+            PoolRegistered regCert
```

Comment:
Is this case correct?
I think this is the only non-obvious about this function, and it's worth a comment.

Reply:
> Is this case correct?

Yes, it's correct. This case represents a re-registration (a registration that occurs after the prior publication of a retirement certificate), in which case the prior retirement is deemed to be obsolete. (See "pool inference rule" from the [Shelley ledger specification](https://hydra.iohk.io/build/3202141/download/1/ledger-spec.pdf))

> I think this is the only non-obvious about this function, and it's worth a comment.

Great point. I've added more comments in fcdfde6ce27c6f7744ecf7728ff1151908d4c959, which also adds a link to the specification.